firebase-firestore/src/main/java/com/google/firebase/firestore/remote/AndroidConnectivityMonitor.java

```diff
@@ -78,34 +86,72 @@ private void configureNetworkMonitoring() {
       final DefaultNetworkCallback defaultNetworkCallback = new DefaultNetworkCallback();
       connectivityManager.registerDefaultNetworkCallback(defaultNetworkCallback);
       unregisterRunnable =
-          new Runnable() {
-            @Override
-            public void run() {
-              connectivityManager.unregisterNetworkCallback(defaultNetworkCallback);
-            }
-          };
+          () -> connectivityManager.unregisterNetworkCallback(defaultNetworkCallback);
     } else {
       NetworkReceiver networkReceiver = new NetworkReceiver();
       @SuppressWarnings("deprecation")
       IntentFilter networkIntentFilter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
       context.registerReceiver(networkReceiver, networkIntentFilter);
-      unregisterRunnable =
-          new Runnable() {
-            @Override
-            public void run() {
-              context.unregisterReceiver(networkReceiver);
-            }
-          };
+      unregisterRunnable = () -> context.unregisterReceiver(networkReceiver);
     }
   }
 
   private void configureBackgroundStateListener() {
-    BackgroundDetector.getInstance().addListener(this);
+    Application applicationContext = (Application) context.getApplicationContext();
+    final AtomicBoolean inBackground = new AtomicBoolean();
+
+    // Manually register an ActivityLifecycleCallback. Android's BackgroundDetector only notifies
+    // when it is certain that the app transitioned from background to foreground. Instead, we
+    // want to be notified whenever there is a slight chance that this transition happened.
+    applicationContext.registerActivityLifecycleCallbacks(
+        new Application.ActivityLifecycleCallbacks() {
+          @Override
+          public void onActivityCreated(@NonNull Activity activity, Bundle savedInstanceState) {}
+
+          @Override
+          public void onActivityStarted(@NonNull Activity activity) {}
+
+          @Override
+          public void onActivityResumed(@NonNull Activity activity) {
+            if (inBackground.compareAndSet(true, false)) {
```

Comment:
Consider putting the body of `onActivityResumed()` into `onActivityCreated()` as well, so that the "offline" state is updated as early as possible (this is what `BackgroundDetector` does). Even copy it into `onActivityStarted()` too.